notebooks/sycamore-tutorial-intermediate-etl.ipynb

```diff
@@ -0,0 +1,410 @@
+    "        raw_date = raw_date.replace(\"Local\", \"\")\n",
+    "        parsed_date = parser.parse(raw_date, fuzzy=True)\n",
+    "        extracted_date = parsed_date.date()\n",
+    "        doc.properties['entity']['day'] = extracted_date.day\n",
+    "        doc.properties['entity']['month'] = extracted_date.month\n",
+    "        doc.properties['entity']['year'] = extracted_date.year\n",
+    "        if parsed_date.utcoffset():\n",
+    "            doc.properties['entity']['dateTime'] = parsed_date.isoformat()\n",
+    "        else:\n",
+    "            doc.properties['entity']['dateTime'] = parsed_date.isoformat() + \"Z\"\n",
+    "    except:\n",
+    "        pass\n",
+    "    return doc\n",
+    "\n",
+    "formatted_docset = chunked_docset.map(convert_timestamp)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's pause and look at what we've done and what's yet to do. So far, we've read a set of PDF documents from S3, partitioned them with the Aryn Partitioning Service, extracted some metadata from them, chunked them into different size elements, and performed some data cleaning.\n",
+    "\n",
+    "What's left is to embed our documents and write them to Pinecone. First, we'll copy the properties from each document to each element of that document using the `spread_properties` transform. Next, First, we will promote these elements to top-level document objects using the `explode` transform. We do this because Sycamore's `embed` transform only operates on documents, not elements."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "exploded_docset = formatted_docset.spread_properties([\"path\", \"entity\"]).explode()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we'll embed each document with the `embed` transform, and generate term frequency tables for sparse vector search with the `term_frequency` transform. Pinecone requires this latter table."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sycamore.transforms.embed import OpenAIEmbedder\n",
+    "\n",
+    "embedded_docset = (exploded_docset\n",
+    "    .embed(OpenAIEmbedder(model_name=embedding_model))\n",
+    "    .term_frequency(tokenizer=tokenizer, with_token_ids=True)\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, we're ready to write our DocSet to Pinecone using Sycamore's Pinecone connector. Because of lazy execution, this will actually execute the code from all of the cells since the last `show`. If you haven't set your Pinecone API key in your env, you will need to specify it in the options in the `write` transform. Also, if your Pinecone database isn't in `us-east-1`, update the region as well."
```

Comment:
This will execute all the code since the last `materialize`. 
You can only get non-IAD if you're paying them. And then you get all kinds of options. Not sure if you want to do anything with that information.

Reply:
Ack